### PR TITLE
3236: Added search field profile filters

### PR DESCRIPTION
**Pull request by cableman**
Extension to field search to allow grouping of profiles and define which should be available on which pages.

See https://platform.dandigbib.org/issues/3236

### CODE REVIEW

File: modules/ting_field_search/plugins/ting_field_search_ctools_export_ui.inc [hunk 1]
```diff
@@ -158,11 +160,61 @@ function ting_field_search_profile_export_form(&$form, &$form_state) {
     '#default_value' => $profile->config['user_interaction']['alter_links'],
   );
 
+  $form['config']['user_interaction']['grouping'] = array(
+    '#type' => 'textfield',
+    '#title' => t('Profile grouping'),
+    '#description' => t('Which group of search profiles should be shown when this profile is used.'),
```
Comment (kasperg) [inline modules/ting_field_search/plugins/ting_field_search_ctools_export_ui.inc]: I do not understand what this description means regarding groups of profiles to be shown when this profile is used. I cannot tell how a profile is added to a group.

File: modules/ting_field_search/plugins/ting_field_search_ctools_export_ui.inc [hunk 2]
```diff
@@ -158,11 +160,61 @@ function ting_field_search_profile_export_form(&$form, &$form_state) {
...
+    '#default_value' => !empty($profile->config['user_interaction']['grouping']) ? $profile->config['user_interaction']['grouping'] : 'default',
+  );
+  $form['config']['user_interaction']['visibility'] = array(
+    '#type' => 'radios',
+    '#title' => t('Show block on specific pages'),
```
Comment (kasperg) [inline modules/ting_field_search/plugins/ting_field_search_ctools_export_ui.inc]: What does "Block" refer to here? Perhaps a wee bit of copy/paste?

File: modules/ting_field_search/ting_field_search.module [hunk 3]
```diff
@@ -432,39 +433,202 @@ function ting_field_search_ting_search_settings_submit($form, &$form_state) {
...
+    }
+
+    $visibility = $profile->config['user_interaction']['visibility'] == 'expect' ? 0 : 1;
+    $pages = $profile->config['user_interaction']['pages'];
+    $path = drupal_strtolower(drupal_get_path_alias($_GET['q']));
+
+    // We have an optimization hack to speed up search (by-pass panels 2x
+    // times loading). So all searches are submitted to "/search-blank",
+    // which need to be added to pages to ensure no form validation errors.
+    // If visibility is "expect" it should not be added to the pages list.
```
Comment (kasperg) [inline modules/ting_field_search/ting_field_search.module]: :+1:

File: modules/ting_field_search/plugins/ting_field_search_ctools_export_ui.inc [hunk 4]
```diff
@@ -158,11 +160,61 @@ function ting_field_search_profile_export_form(&$form, &$form_state) {
...
+    '#options' => array(
+      'all' => t('Show on all pages.'),
+      'expect' => t('All pages except those listed'),
```
Comment (kasperg) [inline modules/ting_field_search/plugins/ting_field_search_ctools_export_ui.inc]: `expect` = `except` - nitpicking but since this is used as a key value later on it should be addressed. Consider constants instead.

File: modules/ting_field_search/ting_field_search.module [hunk 5]
```diff
@@ -432,39 +433,202 @@ function ting_field_search_ting_search_settings_submit($form, &$form_state) {
...
  */
-function ting_field_search_form_search_block_form_alter(&$form, &$form_state) {
-  if (!variable_get('ting_field_search_search_style', FALSE)) {
-    return;
+function ting_field_search_preprocess_html(&$vars) {
+  $profiles = ting_field_search_profiles_load_all();
+
+  foreach ($profiles as $profile) {
+    // Expended search box not checked go to the next profile.
+    if (!array_key_exists('search_box', $profile->config['user_interaction']) || !$profile->config['user_interaction']['search_box']) {
```
Comment (kasperg) [inline modules/ting_field_search/ting_field_search.module]: Could this be replaced with 

```php
if (!empty($profile->config['user_interaction']['search_box'])) {
```

File: modules/ting_field_search/ting_field_search.module [hunk 6]
```diff
@@ -432,39 +433,202 @@ function ting_field_search_ting_search_settings_submit($form, &$form_state) {
...
+  // If there's an active profile and it is stick; set it as default value.
+  if ($default_profile = ting_field_search_get_active_profile()) {
+    $default_profile = $default_profile->config['user_interaction']['sticky'] ? $default_profile->name : '';
+  }
+
+  if (empty($default_profile)) {
+    $profiles = ting_field_search_profiles_load_all();
+    foreach ($profiles as $profile) {
+      // Expended search box not checked go to the next profile.
+      if (!array_key_exists('default_profile', $profile->config['user_interaction']) || !$profile->config['user_interaction']['default_profile']) {
```
Comment (kasperg) [inline modules/ting_field_search/ting_field_search.module]: Some copy/paste in the comment? Also see suggestion regarding value check above.

File: modules/ting_field_search/ting_field_search.module [hunk 7]
```diff
@@ -432,39 +433,202 @@ function ting_field_search_ting_search_settings_submit($form, &$form_state) {
...
+        continue;
+      }
+
+      // If the profile has visibility all always expand the search field.
+      if ($profile->config['user_interaction']['visibility'] === 'all') {
```
Comment (kasperg) [inline modules/ting_field_search/ting_field_search.module]: Do we need to check for the existence of the `visibility` entry here?

File: modules/ting_field_search/ting_field_search.module [hunk 8]
```diff
@@ -432,39 +433,202 @@ function ting_field_search_ting_search_settings_submit($form, &$form_state) {
...
+      $page_match = drupal_match_path($path, $pages);
+      if ($path != $_GET['q']) {
+        $page_match = $page_match || drupal_match_path($_GET['q'], $pages);
+      }
+
+      // When $visibility has a value of 0, the profile is displayed on
+      // all pages except those listed in $pages. When set to 1, it
+      // is displayed only on those pages listed in $pages.
+      if (!($visibility xor $page_match)) {
+        $default_profile = $profile->name;
```
Comment (kasperg) [inline modules/ting_field_search/ting_field_search.module]: This seems to be almost the same code as in `ting_field_search_preprocess_html()`. It is also quite complex. Please refactor.

File: modules/ting_field_search/ting_field_search.module [hunk 9]
```diff
@@ -432,39 +433,202 @@ function ting_field_search_ting_search_settings_submit($form, &$form_state) {
...
+        // If visibility is "expect" it should not be added to the pages list.
+        if ($visibility) {
+          $pages .= "\r\nsearch-blank";
+        }
+
+        // Compare the lowercase internal and lowercase path alias (if any).
+        $page_match = drupal_match_path($path, $pages);
+        if ($path != $_GET['q']) {
+          $page_match = $page_match || drupal_match_path($_GET['q'], $pages);
+        }
```
Comment (kasperg) [inline modules/ting_field_search/ting_field_search.module]: Less copy/paste - more filtering.

File: modules/ting_field_search/ting_field_search.module [hunk 10]
```diff
@@ -618,29 +782,33 @@ function ting_field_search_profiles_load($name) {
...
  *
- * @return array $profiles.
+ * @return array
  *   An array of profile objects keyed bt profile machine name.
  */
 function ting_field_search_profiles_load_all($reset = FALSE) {
-  return ting_field_search_profiles_load_mutliple();
+  $profiles = &drupal_static(__FUNCTION__);
+  if (!isset($profiles) || $reset) {
+    $profiles = ting_field_search_profiles_load_mutliple();
```
Comment (kasperg) [inline modules/ting_field_search/ting_field_search.module]: Typo in `mutliple` but perhaps this matters less here.

File: modules/ting_field_search/ting_field_search.module [hunk 11]
```diff
@@ -823,29 +995,33 @@ function ting_field_search_facet_save($facet) {
...
 /**
  * Delete facet settings associated with a profile.
+ *
+ * @param string $pid
+ *   The id of the profile to delete.
+ * @param string $facet
+ *   Facet to to used as condition for the deletions.
  */
-function ting_field_search_facets_delete($pid, $facet = NULL) {
+function ting_field_search_facets_delete($pid, $facet = '') {
```
Comment (kasperg) [inline modules/ting_field_search/ting_field_search.module]: Why change the default value here?

File: modules/ting_field_search/ting_field_search.ting_field_search_default_profiles.inc [hunk 12]
```diff
@@ -927,6 +957,43 @@ function ting_field_search_ting_field_search_default_profiles() {
...
+    'facets' => array(
+      'use_facets' => 0,
+      'number_of_facets' => 25,
+      'facet_count' => 5,
+      'term_count' => 5,
+    ),
+  );
+  $profile->weight = 11;
+  $profile->facets = array();
+  $export['tegneserier'] = $profile;
```
Comment (kasperg) [inline modules/ting_field_search/ting_field_search.ting_field_search_default_profiles.inc]: Is it intentional that this profile has been added?

File: themes/ddbasic/template.php [hunk 13]
```diff
@@ -34,8 +34,11 @@ function ddbasic_preprocess_html(&$vars) {
       $vars['classes_array'][] = 'search-form-extended';
       $vars['classes_array'][] = 'show-secondary-menu';
 
-      if (menu_get_item()['path'] === 'search/ting/%') {
-        $vars['classes_array'][] = 'extended-search-is-open';
+      $path = menu_get_item()['path'];
+      switch ($path) {
+        case 'search/ting/%';
+        case 'ding_frontpage':
```
Comment (kasperg) [inline themes/ddbasic/template.php]: Adding `ding_frontpage` seems to be a change which is not related to the issue at hand. Is that intentional?

Comment (cableman) [reply themes/ddbasic/template.php]: Yes, as I remember this fixes som issues where you have selected the search field to default be open.

File: modules/ting_field_search/ting_field_search.ting_field_search_default_profiles.inc (hunk 12 above)
Comment (cableman) [reply modules/ting_field_search/ting_field_search.ting_field_search_default_profiles.inc]: It's back in feb. so i really don't remember who request this!

File: modules/ting_field_search/plugins/ting_field_search_ctools_export_ui.inc (hunk 1 above)
Comment (cableman) [reply modules/ting_field_search/plugins/ting_field_search_ctools_export_ui.inc]: Simply grouped by the string/name entered in the field.

File: modules/ting_field_search/plugins/ting_field_search_ctools_export_ui.inc (hunk 2 above)
Comment (cableman) [reply modules/ting_field_search/plugins/ting_field_search_ctools_export_ui.inc]: Fixed :-)

File: modules/ting_field_search/plugins/ting_field_search_ctools_export_ui.inc (hunk 4 above)
Comment (cableman) [reply modules/ting_field_search/plugins/ting_field_search_ctools_export_ui.inc]: Fixed

File: modules/ting_field_search/ting_field_search.module (hunk 5 above)
Comment (cableman) [reply modules/ting_field_search/ting_field_search.module]: It have been changed from your suggestion, based on internal code review to the current form.

File: modules/ting_field_search/ting_field_search.module (hunk 6 above)
Comment (cableman) [reply modules/ting_field_search/ting_field_search.module]: The comment from above her as well.

File: modules/ting_field_search/ting_field_search.module (hunk 8 above)
Comment (cableman) [reply modules/ting_field_search/ting_field_search.module]: This is taken from drupal core, so think this is fine.

File: modules/ting_field_search/ting_field_search.module (hunk 9 above)
Comment (cableman) [reply modules/ting_field_search/ting_field_search.module]: ?

File: modules/ting_field_search/ting_field_search.module (hunk 10 above)
Comment (cableman) [reply modules/ting_field_search/ting_field_search.module]: Fixed

File: modules/ting_field_search/ting_field_search.module (hunk 11 above)
Comment (cableman) [reply modules/ting_field_search/ting_field_search.module]: As all facets are strings, so having NULL as default don't makes sens. The function documentation declare $facet a string.

File: modules/ting_field_search/ting_field_search.module (hunk 7 above)
Comment (cableman) [reply modules/ting_field_search/ting_field_search.module]: I don't remeber why, but a lot of testing (by GBA) has gone into this and there was a lot og issue with frontpage vs. other pages etc....

File: modules/ting_field_search/ting_field_search.module (hunk 8 above)
Comment (kasperg) [reply modules/ting_field_search/ting_field_search.module]: As far as I can tell [the Block module in Drupal Core where this is taken from has this once](https://github.com/drupal/drupal/blob/7.x/modules/block/block.module#L819-L831) and it does not include an optimization hack requiring 4 lines of comments so I do not think it is the same.

See aakbcms/patches#1 for a suggestion for how to refactor.

File: modules/ting_field_search/ting_field_search.module (hunk 9 above)
Comment (kasperg) [reply modules/ting_field_search/ting_field_search.module]: My comment refers to the situation above.

File: modules/ting_field_search/ting_field_search.module (hunk 8 above)
Comment (cableman) [reply modules/ting_field_search/ting_field_search.module]: Changes accepted
